llm/training/finetune.py

```diff
@@ -0,0 +1,74 @@
+import os
+from typing import Dict, Any
+from os.path import join
+import pathlib
+
+# import experiment tracking imports before ML toolkits
+import llm.experiment_tracking_imports  # noqa
+import click
+from transformers import AutoModelForCausalLM, AutoTokenizer, default_data_collator, Trainer
+from peft import prepare_model_for_kbit_training, get_peft_model
+from ruamel.yaml import YAML
+
+from llm.training.config import FineTuneConfig, CopyToSourcesCallback
+
+
+WORLD_SIZE = int(os.getenv("WORLD_SIZE", 1))
+LOCAL_RANK = int(os.getenv("LOCAL_RANK", 0))
+GLOBAL_RANK = int(os.getenv("RANK", 0))
+
+
+def has_checkpoint(dir: str) -> bool:
+    if list(pathlib.Path(dir).glob("checkpoint-*")):
+        return True
+    return False
+
+
+@click.command()
+@click.argument("config_path")
+def run(config_path: str):
+    with open(config_path) as f:
+        config = YAML().load(f)
+    _run(config)
+
+
+def _run(config: Dict[str, Any]):
+    finetune_config = FineTuneConfig.from_config(**config)
+
+    tokenizer = AutoTokenizer.from_pretrained(finetune_config.base_model)
+    device_map = {"": LOCAL_RANK}
+    model = AutoModelForCausalLM.from_pretrained(
+        finetune_config.base_model,
+        load_in_8bit=finetune_config.load_in_8bit,
+        load_in_4bit=finetune_config.load_in_4bit,
+        torch_dtype=finetune_config.torch_dtype,
+        device_map=device_map,
+    )
+
+    model.train()
+    model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=True)
```

Comment:
The fastchat training scripts gate this behind a `q_lora` flag
https://github.com/lm-sys/FastChat/blob/cd7d04879f2f23615fd3e16c4e7c6ccedc8d78c9/fastchat/train/train_lora.py#L156-L159


So probably we'd want to do something similar based on values we use to control quantization_config.